File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int>mp;
        int i = 0, j = 0, n = s.size(), len = 0;
        while(j < n){
            if(mp.count(s[j]) > 0){
                len = max(len, j-i);
                i = max(i, mp[s[j]]+1);
            }
            mp[s[j]] = j;
            j++;
        }
        len = max(len, j-i);
        return len;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_tn, min_tn, max_prod = nums[0], nums[0], nums[0]
        for cur_num in nums[1:]:
            tmp_max_tn = max(cur_num, max_tn * cur_num, min_tn * cur_num)
            min_tn = min(cur_num, max_tn * cur_num, min_tn * cur_num)
            max_tn = tmp_max_tn

            max_prod = max(max_prod, max_tn)
        return max_prod
        
